DOCSUP-1585: Edit and sync geo functions (#12584)
ClickHouse/ClickHouse: docs/en/sql-reference/functions/geo.md

└─────┘
```

> ## Functions for Working with Geohash {#geohash}

Reviewer: we could split all this into:
```
geo/index.md
geo/geohash.md
geo/h3.md
```

Reviewer 2: We will do it in this PR.

> Geohash is the geocode system, which subdivides Earth’s surface into buckets of grid shape and encodes each cell into a short string of letters and digits. It is a hierarchical data structure, so the longer is the geohash string, the more precise is the geographic location.  
> 
> The description of the Geohash system is available in the [Wikipedia](https://en.wikipedia.org/wiki/Geohash). 
> 
> If you need to convert geographic coordinates to geohash strings, you can use [geohash.org](http://geohash.org/).

Reviewer: ```suggestion
[Geohash](https://en.wikipedia.org/wiki/Geohash) is the geocode system, which subdivides Earth’s surface into buckets of grid shape and encodes each cell into a short string of letters and digits. It is a hierarchical data structure, so the longer is the geohash string, the more precise is the geographic location. 

If you need to manually convert geographic coordinates to geohash strings, you can use [geohash.org](http://geohash.org/).
```

└─────────────────────────────────┘
```

### geohashesInBox {#geohashesinbox}

Returns an array of [geohash](#geohash)-encoded strings of given precision that fall inside and intersect boundaries of given box, basically a 2D grid flattened into array.

**Syntax**

``` sql
geohashesInBox(longitude_min, latitude_min, longitude_max, latitude_max, precision)
```

**Parameters**

-   `longitude_min` — Minimum longitude. Range: `[-180°, 180°]`. Type: [Float](../../sql-reference/data-types/float.md).
-   `latitude_min` — Minimum latitude. Range: `[-90°, 90°]`. Type: [Float](../../sql-reference/data-types/float.md).
-   `longitude_max` — Maximum longitude. Range: `[-180°, 180°]`. Type: [Float](../../sql-reference/data-types/float.md).
-   `latitude_max` — Maximum latitude. Range: `[-90°, 90°]`. Type: [Float](../../sql-reference/data-types/float.md).
-   `precision` — Geohash precision. Range: `[1, 12]`. Type: [UInt8](../../sql-reference/data-types/int-uint.md).

!!! info "Note"

>     All coordinate parameters should be of the same type: either `Float32` or `Float64`.

Reviewer 2: ```suggestion
    All coordinate parameters must be of the same type: either `Float32` or `Float64`.
```

**Returned values**

-   Array of precision-long strings of geohash-boxes covering provided area, you should not rely on order of items.

> -   \[\] - Empty array if *min* values of *latitude* and *longitude* aren’t less than corresponding *max* values.

Reviewer 2: ```suggestion
-   `[]` - Empty array if minimum latitude and longitude values aren’t less than corresponding maximum values.
```

Type: [Array](../../sql-reference/data-types/array.md)([String](../../sql-reference/data-types/string.md)).

!!! info "Note"

>     Function will throw an exception if resulting array is over 10’000’000 items long.

Reviewer 2: ```suggestion
    Function throws an exception if resulting array is over 10’000’000 items long.
```

**Example**

Query:

``` sql
SELECT geohashesInBox(24.48, 40.56, 24.785, 40.81, 4) AS thasos
```
Result:

``` text
┌─thasos──────────────────────────────────────┐
│ ['sx1q','sx1r','sx32','sx1w','sx1x','sx38'] │
└─────────────────────────────────────────────┘
```

## Functions for Working with H3 Indexes {#h3_index}

[H3](https://eng.uber.com/h3/) is a geographical indexing system where Earth’s surface divided into a grid of even hexagonal cells. This system is hierarchical, i. e. each hexagon on the top level ("parent") can be splitted into seven even but smaller ones ("children"), and so on. 

The level of the hierarchy is called `resolution` and can receive a value from `0` till `15`, where `0` is the `base` level with the largest and coarsest cells. 

A latitude and longitude pair can be transformed to a 64-bit H3 index, identifying a grid cell.

The H3 index is used primarily for bucketing locations and other geospatial manipulations.

> The full description of the H3 systen is available at [the Uber Engeneering site](https://eng.uber.com/h3/). 

Reviewer 2: ```suggestion
The full description of the H3 system is available at [the Uber Engeneering site](https://eng.uber.com/h3/). 
```
